Update CSS random() function implementation to match the latest spec changes

weinig opened the pull request:
#### a998150461f07d66c785c5aa97bc5c5aedd47bf9
<pre>
Update CSS random() function implementation to match the latest spec changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=291092">https://bugs.webkit.org/show_bug.cgi?id=291092</a>

Reviewed by Antti Koivisto.

Updates the implementation of the CSS random() function to match the latest
spec (<a href="https://drafts.csswg.org/css-values-5/#random).">https://drafts.csswg.org/css-values-5/#random).</a> This includes:

  - New sharing/caching semantics, including replacing the `per-element` keyword with a `match-element` keyword.
  - Removal of the `by` keyword for steps.
  - Updated rules for argument ranges and function execution.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/calc/CSSCalcRandomCachingKey.h: Renamed from Source/WebCore/platform/calc/CalculationRandomKey.h.
* Source/WebCore/css/calc/CSSCalcRandomCachingKeyMap.h: Renamed from Source/WebCore/platform/calc/CalculationRandomKeyMap.h.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserState.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
* Source/WebCore/platform/calc/CalculationExecutor.h:
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Traversal.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleSharingResolver.cpp:

Canonical link: <a href="https://commits.webkit.org/293439@main">https://commits.webkit.org/293439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02727183d7c89d6e91746859963c39141b16f5b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98897 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104023 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100942 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75294 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32416 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101901 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14308 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89312 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55655 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14099 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7284 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48866 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84036 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7357 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106392 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25994 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18949 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84251 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26369 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85510 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83749 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21235 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28406 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6077 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19716 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25947 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31133 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25767 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29087 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
```diff
@@ -628,61 +626,131 @@ static std::optional<TypedChild> consumeRound(CSSParserTokenRange& tokens, int d
+static std::optional<Random::SharingOptions> consumeOptionalRandomSharingOptions(CSSParserTokenRange& tokens, ParserState& state)
 {
-    // <random-caching-options> = <dashed-ident> || per-element
+    // <random-value-sharing-options> = [ [ auto | <dashed-ident> ] || match-element ]
```
I think the issue went with `element-shared`: https://github.com/w3c/csswg-drafts/issues/11742#issuecomment-2755039081

weinig replied:
We can change it if/when they update the spec.

nt1m replied:
element-shared is the name kept by the resolution, I would go with that instead of match-element which was never resolved against. Maybe @fantasai can confirm

weinig committed:
Update CSS random() function implementation to match the latest spec changes
https://bugs.webkit.org/show_bug.cgi?id=291092

Reviewed by Antti Koivisto.

Updates the implementation of the CSS random() function to match the latest
spec (https://drafts.csswg.org/css-values-5/#random). This includes:

  - New sharing/caching semantics, including replacing the `per-element` keyword with a `match-element` keyword.
  - Removal of the `by` keyword for steps.
  - Updated rules for argument ranges and function execution.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/random-serialize.tentative.html:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/calc/CSSCalcRandomCachingKey.h: Renamed from Source/WebCore/platform/calc/CalculationRandomKey.h.
* Source/WebCore/css/calc/CSSCalcRandomCachingKeyMap.h: Renamed from Source/WebCore/platform/calc/CalculationRandomKeyMap.h.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserState.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
* Source/WebCore/platform/calc/CalculationExecutor.h:
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp:
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp:
* Source/WebCore/platform/calc/CalculationTree+Traversal.h:
* Source/WebCore/platform/calc/CalculationTree.cpp:
* Source/WebCore/platform/calc/CalculationTree.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleSharingResolver.cpp:

Canonical link: https://commits.webkit.org/293439@main
